examples/model-viewer/main.cpp

```diff
@@ -0,0 +1,922 @@
+            model->indexCount = data.indexCount;
+
+            int meshCount = data.meshCount;
+            for (int ii = 0; ii < meshCount; ++ii)
+                model->meshes.push_back((Mesh*)data.meshes[ii]);
+
+            return model;
+        }
+    };
+    Callbacks callbacks;
+    callbacks.context = context;
+
+    // We instantiate a model loader object and then use it to
+    // try and load a model from the chosen path.
+    //
+    platform::ModelLoader loader;
+    loader.device = context->device;
+    loader.loadFlags = loadFlags;
+    loader.scale = scale;
+    loader.callbacks = &callbacks;
+    Model* model = nullptr;
+    if (SLANG_FAILED(loader.load(inputPath, (void**)&model)))
+    {
+        log("failed to load '%s'\n", inputPath);
+        return nullptr;
+    }
+
+    return model;
+}
+
+// Along with materials, our application needs to be able to represent
+// multiple light sources in the scene. For this task we will use a C++
+// inheritance hierarchy rooted at `Light` to match the `ILight`
+// interface in Slang.
+
+struct Light : RefObject
+{
+    // A light must be able to create a shader object defining its
+    // corresponding shader type and parameter values.
+    virtual IShaderObject* createShaderObject(RendererContext* context) = 0;
+
+    // Retrieves the shader type for this light object.
+    virtual slang::TypeReflection* getShaderType(RendererContext* context) = 0;
+
+    // The shader object for a light will be stashed here
+    // after it is created.
+    ComPtr<IShaderObject> shaderObject;
+};
+
+// Helper function to retrieve the underlying shader type of `T`.
+template<typename T>
+slang::TypeReflection* getShaderType(RendererContext* context)
+{
+    auto program = context->slangReflection;
+    auto shaderType = program->findTypeByName(T::getTypeName());
+    return shaderType;
+}
+
+// We will provide two nearly trivial implementations of `Light` for now,
+// to show the kind of application code needed to line up with the corresponding
+// types defined in the Slang shader code for this application.
+
+struct DirectionalLight : Light
+{
+    glm::vec3 direction = normalize(glm::vec3(1));
+    glm::vec3 intensity = glm::vec3(1);
+
+    static const char* getTypeName() { return "DirectionalLight"; }
+
+    virtual IShaderObject* createShaderObject(RendererContext* context) override
+    {
+        auto shaderType = ::getShaderType<DirectionalLight>(context);
+        shaderObject = context->device->createShaderObject(shaderType);
+        gfx::ShaderCursor cursor(shaderObject);
+        cursor["direction"].setData(&direction, sizeof(direction));
+        cursor["intensity"].setData(&intensity, sizeof(intensity));
+        return shaderObject.get();
+    }
+
+    virtual slang::TypeReflection* getShaderType(RendererContext* context) override
+    {
+        return ::getShaderType<DirectionalLight>(context);
+    }
+};
+
+struct PointLight : Light
```

Comment:
It's kind of an obvious statement, but if/when we start to support more complete heterogeneous programs, we will need to decide on the best way to support types like this where there is a host side `X` and a GPU-side `X` that have basically the same members/data. It seems clear that we don't want to have to duplicate the declarations here and deal with manually copying from CPU to GPU, but there are also many reasons why it can be useful to have distinct host and GPU-memory allocations for a shader object, and to make creating one from the other explicit. This needs careful thought as we continue to iterate on these examples.